mm2src/mm2_metrics/src/recorder.rs

```diff
@@ -0,0 +1,231 @@
+    //         output.push('\n');
+    //     }
+
+    //     output
+    // }
+
+    pub fn prepare_json(&self) -> MetricsJson {
+        let mut output = vec![];
+
+        for counters in self.registry.get_counter_handles() {
+            let (key, labels) = counters.0.into_parts();
+            let value = counters.1.get_inner().load(Ordering::Acquire);
+            output.push(MetricType::Counter {
+                key: key.as_str().to_string(),
+                labels: labels_into_parts(labels.clone().iter()),
+                value,
+            });
+        }
+
+        for counters in self.registry.get_gauge_handles() {
+            let (key, labels) = counters.0.into_parts();
+            let value = f64::from_bits(counters.1.get_inner().load(Ordering::Acquire));
+            output.push(MetricType::Gauge {
+                key: key.as_str().to_string(),
+                labels: labels_into_parts(labels.clone().iter()),
+                value,
+            });
+        }
+
+        for (key, value) in self.registry.get_histogram_handles() {
+            let value = value.get_inner().data();
+            let (key, labels) = key.into_parts();
+            let mm_histogram = MmHistogram::new(&value);
+
+            if let Some(qauntiles_value) = mm_histogram {
+                output.push(MetricType::Histogram {
+                    key: key.as_str().to_string(),
+                    labels: labels_into_parts(labels.clone().iter()),
+                    quantiles: qauntiles_value.to_json_quantiles(),
+                });
+            }
+        }
+
+        MetricsJson { metrics: output }
+    }
+}
+
+// fn key_value_to_snapshot_entry<T: Clone>(key: Key, value: T) -> HashMap<Vec<String>, T> {
+//     let (_name, labels) = key_to_parts(&key, None);
+//     let mut entry = HashMap::new();
+//     entry.insert(labels, value);
+//     entry
+// }
+
+impl Recorder for MmRecorder {
+    fn describe_counter(&self, _key_name: KeyName, _unit: Option<Unit>, _description: &'static str) {
+        // mm2_metrics doesn't use this method
+    }
+
+    fn describe_gauge(&self, _key_name: KeyName, _unit: Option<Unit>, _description: &'static str) {
+        // mm2_metrics doesn't use this method
+    }
+
+    fn describe_histogram(&self, _key_name: KeyName, _unit: Option<Unit>, _description: &'static str) {
+        // mm2_metrics doesn't use this method
+    }
+
+    fn register_counter(&self, key: &Key) -> Counter { self.registry.get_or_create_counter(key, |e| e.clone().into()) }
+
+    fn register_gauge(&self, key: &Key) -> Gauge { self.registry.get_or_create_gauge(key, |e| e.clone().into()) }
+
+    fn register_histogram(&self, key: &Key) -> metrics::Histogram {
+        self.registry.get_or_create_histogram(key, |e| e.clone().into())
+    }
+}
+
+pub trait TryRecorder {
+    /// Check for recorder and set one if none is set.
+    fn try_recorder(&self) -> Option<Arc<MmRecorder>>;
+}
+
+pub(crate) fn labels_to_tags(labels: Iter<Label>) -> Vec<Tag> {
```

Comment:
I thing it's better to move this function to the module where the `TagObserver` is.